Timelock/project/timelock-contract/src/main.sw

```diff
@@ -0,0 +1,135 @@
+contract;
+
+dep data_structures;
+dep errors;
+dep events;
+dep interface;
+dep utils;
+
+use data_structures::ExecutionRange;
+use errors::{AccessControlError, FundingError, TransactionError};
+use events::{CancelEvent, ExecuteEvent, QueueEvent};
+use interface::{Info, Timelock};
+use std::{
+    auth::msg_sender,
+    block::timestamp as now,
+    bytes::Bytes,
+    call_frames::msg_asset_id,
+    context::this_balance,
+    logging::log,
+};
+use utils::create_hash;
+
+const ADMIN: Identity = Identity::Address(Address::from(OWNER));
+
+storage {
+    /// Mapping transaction hash to time range of available execution
+    queue: StorageMap<b256, Option<ExecutionRange>> = StorageMap {},
+}
+
+impl Timelock for Contract {
+    #[storage(read, write)]
+    fn cancel(id: b256) {
+        require(msg_sender().unwrap() == ADMIN, AccessControlError::AuthorizationError);
+        require(storage.queue.get(id).is_some(), TransactionError::InvalidTransaction(id));
+
+        storage.queue.insert(id, Option::None::<ExecutionRange>());
+
+        log(CancelEvent { id })
+    }
+
+    #[storage(read, write)]
+    fn execute(
+        recipient: Identity,
+        value: Option<u64>,
+        asset_id: Option<ContractId>,
+        data: Bytes,
+        timestamp: u64,
+    ) {
+        require(msg_sender().unwrap() == ADMIN, AccessControlError::AuthorizationError);
+
+        let id = create_hash(recipient, value, asset_id, data, timestamp);
+        let transaction = storage.queue.get(id);
+
+        require(transaction.is_some(), TransactionError::InvalidTransaction(id));
+
+        // Timestamp is guarenteed to be in the range because of `fn queue()`
+        // Therefore, the lower bound can be the timestamp itself; but, we must place an upper bound
+        // to prevent going over the MAXIMUM_DELAY
+        require(timestamp <= now() && now() <= transaction.unwrap().end, TransactionError::TimestampNotInRange((timestamp, transaction.unwrap().end, now())));
```

Comment:
So I thought the purpose of the timelock contract was to be able to execute a transaction after some timestamp.  Here the execution has an upper bound of `MAXIMUM_DELAY`.  I thought the delay represented a duration of time until the queued transaction could be executed and not some upper bound that creates an execution window.  I'm fine with having an execution window, but I don't think it makes sense having the window's upper bound be`MAXIMUM_DELAY`.  IMO the upper bound should be either: a new config time constant, a parameter you pass in to queue, or non-existant (have no upper bound on timelock).

Reply:
The upper bound is already a config time constant so I don't understand how having another one makes a difference.

It makes sense to me that there is an expiration time for the transaction. You may not want to allow an indefinite transfer window. As you've pointed out this could be an optional argument.
The logic would need to be adjusted a bit.

The assumption was that the contract has a set window and if you want a different contract window then you dedicate a new contract which differentiates between one contract and another. Is that needed? Maybe, maybe not. It's a design decision which I don't think really matters to us. I could adjust it to be more flexible by passing in an optional maximum delay if you'd like. Can you confirm on which action to take?